### PR TITLE
Feature/BDN-643: Merge demand and bidding to one auction

### PR DESCRIPTION
**Pull request by AntonZarytski**
https://appodeal.atlassian.net/browse/BDN-643

### CODE REVIEW

File: adapter/amazon/src/main/java/org/bidon/amazon/ext/Ext.kt [hunk 1]
```diff
@@ -2,6 +2,22 @@ package org.bidon.amazon.ext
 
 import com.amazon.device.ads.AdRegistration
 import org.bidon.amazon.BuildConfig
+import org.bidon.amazon.SlotType
+import org.json.JSONArray
 
 internal var adapterVersion = BuildConfig.ADAPTER_VERSION
 internal var sdkVersion = AdRegistration.getVersion()
+
+internal fun JSONArray.toSlots(): Map<SlotType, List<String>> {
```
Comment (da2gl) [inline adapter/amazon/src/main/java/org/bidon/amazon/ext/Ext.kt]: why? Do you have already `ParseSlotsUseCase`

File: adapter/amazon/src/main/java/org/bidon/amazon/AmazonAdapter.kt [hunk 2]
```diff
@@ -76,4 +79,8 @@ class AmazonAdapter :
     override fun rewarded(): AdSource.Rewarded<FullscreenAuctionParams> {
         return AmazonRewardedImpl()
     }
+
+    companion object {
+        var slots: Map<SlotType, List<String>> = mapOf()
```
Comment (da2gl) [inline adapter/amazon/src/main/java/org/bidon/amazon/AmazonAdapter.kt]: u need just pickup it from main

File: adapter/amazon/src/main/java/org/bidon/amazon/impl/AmazonBannerImpl.kt [hunk 3]
```diff
@@ -43,7 +45,9 @@ internal class AmazonBannerImpl :
         get() = adView != null
 
     override suspend fun getToken(context: Context, adTypeParam: AdTypeParam, adUnits: List<AdUnit>): String? {
-        val slots = ParseSlotsUseCase()(adUnits).also {
+        val slots = AmazonAdapter.slots.filter {
```
Comment (da2gl) [inline adapter/amazon/src/main/java/org/bidon/amazon/impl/AmazonBannerImpl.kt]: it already filtered in ObtainTokenUseCase

File: bidon/src/main/java/org/bidon/sdk/ads/banner/BannerView.kt [hunk 4]
```diff
@@ -44,6 +44,7 @@ class BannerView @JvmOverloads constructor(
     context: Context,
     attrs: AttributeSet? = null,
     @AttrRes defStyleAtt: Int = 0,
+    private val auctionKey : String? = null,
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/ads/banner/BannerView.kt]: why private?

File: bidon/src/main/java/org/bidon/sdk/ads/interstitial/InterstitialAd.kt [hunk 5]
```diff
@@ -8,7 +8,7 @@ import org.bidon.sdk.stats.WinLossNotifier
 /**
  * Created by Bidon Team on 06/02/2023.
  */
-class InterstitialAd : Interstitial by InterstitialImpl()
+class InterstitialAd(auctionKey: String? = null) : Interstitial by InterstitialImpl(auctionKey = auctionKey)
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/ads/interstitial/InterstitialAd.kt]: `@JvmOverloads constructor`

File: bidon/src/main/java/org/bidon/sdk/ads/rewarded/RewardedAd.kt [hunk 6]
```diff
@@ -8,7 +8,7 @@ import org.bidon.sdk.stats.WinLossNotifier
 /**
  * Created by Bidon Team on 06/02/2023.
  */
-class RewardedAd : Rewarded by RewardedImpl()
+class RewardedAd(auctionKey: String? = null) : Rewarded by RewardedImpl(auctionKey = auctionKey)
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/ads/rewarded/RewardedAd.kt]: `@JvmOverloads constructor`

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 7]
```diff
@@ -76,15 +91,18 @@ internal class AuctionImpl(
...
                         adapters = adaptersSource.adapters.associate {
                             it.demandId.demandId to it.adapterInfo
-                        }
+                        },
+                        tokens = tokens,
                     ).mapCatching { auctionData ->
-                        if (auctionId != auctionData.auctionId) {
-                            logError(TAG, "Auction ID has been changed", IllegalStateException())
-                        }
+                        //TODO uncomment after mock tests
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: TODO

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 8]
```diff
@@ -76,15 +91,18 @@ internal class AuctionImpl(
...
+//                      if (auctionId != auctionData.auctionId) {
+//                          logError(TAG, "Auction ID has been changed", IllegalStateException())
+//                      }
                         conductAuction(
                             auctionData = auctionData,
                             demandAd = demandAd,
                             adTypeParamData = adTypeParam,
+                            tokens = tokens,
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt:103]: tokens here never use

File: bidon/src/main/java/org/bidon/sdk/auction/models/AdUnit.kt [hunk 9]
```diff
@@ -12,7 +12,8 @@ data class AdUnit(
     val pricefloor: Double?,
     val uid: String,
     val bidType: BidType,
-    private val ext: String?,
+    val payload: String?,
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/models/AdUnit.kt]: `payload` gets every ad network itself if it needs it

Comment (da2gl) [reply bidon/src/main/java/org/bidon/sdk/auction/models/AdUnit.kt]: image

File: bidon/src/main/java/org/bidon/sdk/auction/models/AuctionResponse.kt [hunk 10]
```diff
@@ -22,14 +22,18 @@ internal class AuctionResponseParser : JsonParser<AuctionResponse> {
...
             auctionId = json.getString("auction_id"),
+            auctionTimeout = json.optLong("auction_timeout", defaultTimeout),
             auctionConfigurationId = json.optLong("auction_configuration_id"),
             auctionConfigurationUid = json.optString("auction_configuration_uid"),
             externalWinNotificationsEnabled = json.optBoolean("external_win_notifications", false),
         )
     }.getOrNull()
 }
+
+private const val defaultTimeout = 30000L
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/models/AuctionResponse.kt]: TODO

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 11]
```diff
@@ -65,8 +67,21 @@ internal class AuctionImpl(
             }
             this.adTypeParam = adTypeParam
             job = scope.launch {
+                resultsCollector.startRound(adTypeParam.pricefloor)
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: may be move in runCatching ?

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 12]
```diff
@@ -65,8 +67,21 @@ internal class AuctionImpl(
...
+
+                val auctionId = UUID.randomUUID().toString()
+                resultsCollector.serverBiddingStarted()
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: serverBiddingStarted used for what?

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 13]
```diff
@@ -65,8 +67,21 @@ internal class AuctionImpl(
...
+
+                val tokens = get<GetTokensUseCase>().invoke(
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: move to constructor or not?

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt [hunk 14]
```diff
@@ -65,8 +67,21 @@ internal class AuctionImpl(
...
+                    adType = demandAd.adType,
+                    adTypeParam = adTypeParam,
+                    adaptersSource = adaptersSource
+                )
+                tokens.forEachIndexed { index, (demandId, token) ->
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: printed twice here and GetTokensUseCaseImpl

File: bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+         * Tokens Obtaining
+         */
+        val tokens = biddingAdSources.getTokens(
+            context = get(),
+            adTypeParam = adTypeParam,
+        )
+        logInfo(TAG, "${tokens.size} token(s):")
+        tokens.forEachIndexed { index, (demandId, token) ->
+            logInfo(TAG, "#$index $demandId {$token}")
+        }
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt]:
```
if (tokens.all { it.second.status != TokenInfo.Status.SUCCESS.code }) {
                    logError(TAG, "No tokens found", BidonError.NoBid)
                    resultsCollector.serverBiddingFinished(null)
                    return@withTimeoutOrNull
                }
```
lose this logic

File: bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+                    adTypeParam = adTypeParam,
+                )
+            }
+            .filterIsInstance<Mode.Bidding>()
+
+        /**
+         * Tokens Obtaining
+         */
+        val tokens = biddingAdSources.getTokens(
+            context = get(),
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt]: context = adTypeParam.activity.applicationContext,
see past implementation

File: bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+                        "GDPR=${BidonSdk.regulation.gdpr}, " +
+                        "COPPA=${BidonSdk.regulation.coppa}, " +
+                        "usPrivacyString=${BidonSdk.regulation.usPrivacyString}, " +
+                        "gdprConsentString=${BidonSdk.regulation.gdprConsentString}"
+            )
+            supportsRegulation.updateRegulation(BidonSdk.regulation)
+        }
+    }
+
+    private fun List<Adapter>.getAdSources(adType: AdType): List<AdSource<AdAuctionParams>> =
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt]: Set<Adapter> and delete toList unnecessary cast above

File: bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+    ): List<Pair<String, TokenInfo>> {
+        /**
+         * Bidding demands auction
+         */
+        val filteredBiddingAdapters =
+            adaptersSource.adapters.onEach(::applyRegulation).toList()
+        val biddingAdSources = filteredBiddingAdapters
+            .getAdSources(adType)
+            .onEach {
+                applyParams(
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/GetTokensUseCaseImpl.kt]: do we really need this applyParams?

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt (hunk 12 above)
Comment (AntonZarytski) [reply bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: It needs to set timestamp of serverBidding started(when we start to receiving tokens)

File: bidon/src/main/java/org/bidon/sdk/auction/models/AdUnit.kt [hunk 19]
```diff
@@ -9,10 +9,10 @@ import org.json.JSONObject
 data class AdUnit(
     val demandId: String,
     val label: String,
-    val pricefloor: Double?,
+    val pricefloor: Double,
     val uid: String,
     val bidType: BidType,
-    private val ext: String?,
+    val ext: String?,
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/auction/models/AdUnit.kt]: private

File: bidon/src/main/java/org/bidon/sdk/config/models/BiddingConfig.kt [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+import org.bidon.sdk.utils.json.JsonParser
+import org.json.JSONObject
+
+internal interface BiddingConfig {
+    var tokenTimeout: Long
+}
+
+internal class BiddingConfigImpl(override var tokenTimeout: Long = 20000): BiddingConfig
+
+internal class BiddingResponseParser : JsonParser<BiddingConfig> {
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/config/models/BiddingConfig.kt]: setup to JsonParsers.kt

File: bidon/src/main/java/org/bidon/sdk/config/models/ConfigResponse.kt [hunk 21]
```diff
@@ -8,19 +8,23 @@ import org.json.JSONObject
...
-            initializationTimeout = json.getLong("tmax"),
-            adapters = json.getJSONObject("adapters").let { jsonAdapters ->
+            initializationTimeout = init.getLong("tmax"),
+            adapters = init.getJSONObject("adapters").let { jsonAdapters ->
                 jsonAdapters.keys().asSequence().associateWith { adapterName ->
                     jsonAdapters.getJSONObject(adapterName)
                 }
-            }
+            },
+            bidding = BiddingResponseParser().parseOrNull(bidding.toString())
```
Comment (da2gl) [inline bidon/src/main/java/org/bidon/sdk/config/models/ConfigResponse.kt]: see parseSegmentUid

File: bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt (hunk 11 above)
Comment (da2gl) [reply bidon/src/main/java/org/bidon/sdk/auction/impl/AuctionImpl.kt]: runCatching?

File: adapter/admob/src/main/java/org/bidon/admob/AdmobInitParameters.kt [hunk 22]
```diff
@@ -39,9 +38,9 @@ sealed interface AdmobBannerAuctionParams : AdAuctionParams {
         override val bannerFormat: BannerFormat,
         override val containerWidth: Float,
         override val price: Double,
-        bidResponse: BidResponse
+        bidResponse: AdUnit
```
Comment (da2gl) [inline adapter/admob/src/main/java/org/bidon/admob/AdmobInitParameters.kt]: bidResponse -> override val adUnit: AdUnit = bidResponse
for all adnetworks in different pr

File: bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/AuctionStatImpl.kt [hunk 23]
```diff
@@ -61,316 +59,188 @@ internal class AuctionStatImpl(
...
+        val results: List<StatsAdUnit> = roundResults
+            .map { it.asStatsAdUnit() }
+            // TODO try to find more useful solution, cause after auction ends, for filled ad we
+            // receive Successful("INTERNAL_STATUS")
+            .map { statsAdUnit ->
+                val currentUuid = statsAdUnit.adUnitUid
+                if (winnerUuid == currentUuid) {
+                    statsAdUnit.copy(
+                        demandId = statsAdUnit.demandId,
+                        status = RoundStatus.Win.code,
```
Comment (AntonZarytski) [inline bidon/src/main/java/org/bidon/sdk/auction/usecases/impl/AuctionStatImpl.kt:89]: in this case, we should change status Successful("INTERNAL_STATUS") to Win("WIN")

File: bidon/src/main/java/org/bidon/sdk/config/impl/BidonErrorExt.kt [hunk 24]
```diff
@@ -6,8 +6,17 @@ import org.bidon.sdk.config.BidonError
...
+        isJobCancellationException(this) -> BidonError.AuctionCancelled
+        else -> BidonError.Unspecified(
+            demandId = null,
+            sourceError = this
+        )
+    }
+}
+
+//TODO try to find more useful solution
+private fun isJobCancellationException(throwable: Throwable): Boolean {
```
Comment (AntonZarytski) [inline bidon/src/main/java/org/bidon/sdk/config/impl/BidonErrorExt.kt:20]: @da2gl Occurs when AdObject onDestroy was called during auction. In this case we should send AuctionCancelled round result

**Pull request merged**